graph/pull.go

```diff
@@ -635,24 +638,38 @@ func (s *TagStore) pullV2Tag(r *registry.Session, out io.Writer, endpoint *regis
 			return nil
 		}
 
+		// Try to lock the image to make sure we're the only one downloading.
+		// Note: If we successfully lock (pullAlreadInProgress = false), then we are
+		// also responsible for removing ourselves from the pool, in success or failure.
+		c, err := s.poolAdd("pull", "img:"+img.ID)
+		downloads[i].pullAlreadyInProgress = err != nil
+		downloads[i].waitChannel = c
+
 		downloads[i].err = make(chan error)
 		go func(di *downloadInfo) {
 			di.err <- downloadFunc(di)
 		}(&downloads[i])
 	}
 
+	var lastErr error
 	var tagUpdated bool
 	for i := len(downloads) - 1; i >= 0; i-- {
 		d := &downloads[i]
 		if d.err != nil {
 			if err := <-d.err; err != nil {
-				return false, err
+				logrus.Debugf("Failed to download %s with err %s", d.img.ID, err)
+				lastErr = err
```

Comment:
Behavior of this loop has been changed; now, we do not exit early if one of the downloads failed, but rather wait for all parallel downloads to fail. This allows us to correctly clean up the temp files and unlock the images.